File: KeywordBidding.py
#IMPORT NECESSARY PACKAGES

import pandas as pd
import numpy as np
import re

#IMPORT DATA

Inventory_Current_Onsite=pd.read_csv("Inventory_Current_Onsite.csv")
Inventory_Historical=pd.read_csv("Inventory_Historical.csv")
KW_Attributes=pd.read_csv("KW_Attributes.csv")
KW_Performance_L120D=pd.read_csv("KW_Performance_L120D.csv")
Make_Model_ASR=pd.read_csv("Make_Model_ASR.csv")

#INSPECT DATA

print("\n INSPECT DATA: \n")
print("\n Inventory_Current_Onsite: \n")
print(Inventory_Current_Onsite.info())
print("\n Inventory_Historical: \n")
print(Inventory_Historical.info())
print("\n KW_Attributes: \n")
print(KW_Attributes.info())
print("\n KW_Performance_L120D: \n")
print(KW_Performance_L120D.info())
print("\n Make_Model_ASR: \n")
print(Make_Model_ASR.info())
print("\n Make_Model_ASR Tail:\n")
print(Make_Model_ASR.tail(92))

''' 
CLEANING NOTES:

* No missing values observed
* One empty column in Inventory_Historical
* Few empty trailing rows in Make_Model_ASR
* Following columns need to be converted to integer/float type after removing charecters like "$" and/or ",":
   Est First Pos. Bid (KW_Attributes)
   Est Top of Page Bid (KW_Attributes)
   Cost (KW_Performance_L120D)
   Impressions (KW_Performance_L120D)
   Clicks (KW_Performance_L120D)
   ASR (Make_Model_ASR)

'''

#CLEAN DATA

Make_Model_ASR=Make_Model_ASR.dropna(how='all')
Inventory_Historical=Inventory_Historical.drop(Inventory_Historical.columns[4],axis=1)
KW_Attributes['Est First Pos. Bid'] = KW_Attributes['Est First Pos. Bid'].apply(lambda x: float(re.sub('\\$|,','',x)))
KW_Attributes['Est Top of Page Bid'] = KW_Attributes['Est Top of Page Bid'].apply(lambda x: float(re.sub('\\$|,','',x)))
Make_Model_ASR['ASR'] = Make_Model_ASR['ASR'].apply(lambda x: float(re.sub('\\$|,', '',x)))
KW_Performance_L120D['Cost'] = KW_Performance_L120D['Cost'].apply(lambda x: float(re.sub('\\$|,','',x)))
KW_Performance_L120D['Impressions'] = KW_Performance_L120D['Impressions'].apply(lambda x: int(re.sub(',','',x)))
KW_Performance_L120D['Clicks'] = KW_Performance_L120D['Clicks'].apply(lambda x: int(re.sub(',', '',x)))


#EXTRACT ATTRIBUTES (MKT, MAKE, MODEL and YEAR FROM AD GROUP)

KW_Attributes['MKT']=KW_Attributes['Ad group'].apply(lambda x: re.split('-|_',x)[2])
KW_Attributes['Make']=KW_Attributes['Ad group'].apply(lambda x: re.split('-|_',x)[4])
KW_Attributes['Model']=KW_Attributes['Ad group'].apply(lambda x: re.split('-|_',x)[6])
KW_Attributes['Year']=KW_Attributes['Ad group'].apply(lambda x: int("20"+re.split('-|_',x)[8]))

#DATA INTEGRITY CHECKS

print("\n DATA INTEGRITY CHECKS: \n")
print("\n Check for Duplicates: \n")
print("\n KW_Attributes has " + str(KW_Attributes[['KW ID']].duplicated().sum())+ " duplicates in KW ID \n")
print("\n KW_Performance_L120D has " + str(KW_Performance_L120D[['KW ID']].duplicated().sum())+ " duplicates in KW ID \n")
print("\n Inventory_Historical has " + str(Inventory_Historical[['Make','Model','Year']].duplicated().sum())+ " duplicates in [Make,Model,Year] \n")
print("\n Inventory_Current_Onsitel has " + str(Inventory_Current_Onsite[['Make','Model','Year']].duplicated().sum())+ " duplicates in [Make,Model,Year] \n")
print("\n Make_Model_ASR has " + str(Make_Model_ASR[['Make','Model']].duplicated().sum())+ " duplicates in [Make,Model] \n")
print("\n Referential Integrity Check: \n")
merged = pd.merge(KW_Attributes,KW_Performance_L120D, on='KW ID',how='outer', indicator=True)
print("\n There are " + str(np.sum(merged['_merge']!='both'))+ " rows between KW_Attributes and KW_Performance_L120D, that do not match when joined on KW ID \n") 
merged = pd.merge(KW_Attributes,Inventory_Historical,on=['Make','Model','Year'],how='outer', indicator=True)
print("\n There are " + str(np.sum(merged['_merge']!='both'))+ " rows between KW_Attributes and Inventory_Historical, that do not match when joined on [Make,Model,Year] \n")
merged = pd.merge(KW_Attributes,Inventory_Current_Onsite,on=['Make','Model','Year'],how='outer', indicator=True)
print("\n There are " + str(np.sum(merged['_merge']!='both'))+ " rows between KW_Attributes and Inventory_Current_Onsite, that do not match when joined on [Make,Model,Year] \n")
merged = pd.merge(KW_Attributes,Make_Model_ASR,on=['Make','Model'],how='outer', indicator=True)
print("\n There are " + str(np.sum(merged['_merge']!='both'))+ " rows between KW_Attributes and Make_Model_ASR, that do not match when joined on [Make,Model] \n")



'''
DATA INTEGRITY CHECKS - SUMMARY:
No duplicates found in unique keys (KW ID) and composite unique keys ([Make,Model,Year] & [Make,Model])
No referential integrity violations observed

'''

#PREPARE DATA

KW_Attributes=KW_Attributes.merge(KW_Performance_L120D,on='KW ID',how='inner')
KW_Attributes['CVR_Key Word']=KW_Attributes['Conversions']/KW_Attributes['Clicks']
KW_Attributes['Conversions_Ad_group']=KW_Attributes.groupby('Ad group')['Conversions'].transform(np.sum)
KW_Attributes['CVR_Ad_group']=KW_Attributes['Conversions_Ad_group']/(KW_Attributes.groupby('Ad group')['Clicks'].transform(np.sum))
KW_Attributes['Conversions_Make_Model_Year']=KW_Attributes.groupby(['Make','Model','Year'])['Conversions'].transform(np.sum)
KW_Attributes['CVR_Make_Model_Year']=KW_Attributes['Conversions_Make_Model_Year']/(KW_Attributes.groupby(['Make','Model','Year'])['Clicks'].transform(np.sum))
KW_Attributes['Conversions_Make_Model']=KW_Attributes.groupby(['Make','Model'])['Conversions'].transform(np.sum)
KW_Attributes['CVR_Make_Model']=KW_Attributes['Conversions_Make_Model']/(KW_Attributes.groupby(['Make','Model'])['Clicks'].transform(np.sum))
KW_Attributes['Conversions_MKT']=KW_Attributes.groupby('MKT')['Conversions'].transform(np.sum)
KW_Attributes['CVR_MKT']=KW_Attributes['Conversions_MKT']/(KW_Attributes.groupby('MKT')['Clicks'].transform(np.sum))
KW_Attributes=KW_Attributes.merge(Inventory_Historical,on=['Make','Model','Year'],how='inner') 
KW_Attributes=KW_Attributes.merge(Inventory_Current_Onsite,on=['Make','Model','Year'],how='inner')
KW_Attributes=KW_Attributes.merge(Make_Model_ASR[['Make','Model','ASR']],on=['Make','Model'],how='inner')
KW_Attributes=KW_Attributes.set_index('KW ID')

#STEP 1- CALCULATE INITIAL BID

def generate_initial_bid (KW):
    """Uses the logic described in the excercsie 
    and returns an initial bid value for each keyword."""    
    if KW['Conversions'] > 10:
        return (KW['CVR_Key Word']*KW['ASR'])
    elif KW['Conversions_Ad_group']>10:
        return (KW['CVR_Ad_group']*KW['ASR'])
    elif KW['Conversions_Make_Model_Year']>10:
        return (KW['CVR_Make_Model_Year']*KW['ASR'])
    elif KW['Conversions_Make_Model']>10:
        return (KW['CVR_Make_Model']*KW['ASR']) 
    else:
        return (KW['Est First Pos. Bid']) 


KW_Attributes['Bid']=KW_Attributes.apply(generate_initial_bid,axis=1)
              
                                       
#STEP 2- ADJUST CALCULATED BID (VECTORIZED IMPLEMENTATION)   

#a) Adjust bid based on current onsite inventory            
                                    
InvAdj=KW_Attributes['CurrentOnsiteInventory']<KW_Attributes['HistAvgInv']
AdjFactor=((KW_Attributes.loc[InvAdj,'HistAvgInv']-KW_Attributes.loc[InvAdj,'CurrentOnsiteInventory'])/KW_Attributes.loc[InvAdj,'HistAvgInv'])/2
KW_Attributes.loc[InvAdj,'Bid']=(1-AdjFactor)*KW_Attributes.loc[InvAdj,'Bid']

#b) Adjust bid based on Mkt CVR

MktAdj=np.logical_and(KW_Attributes['Conversions_Ad_group']<11 ,np.logical_or(KW_Attributes['Conversions_Make_Model_Year']>10,KW_Attributes['Conversions_Make_Model']>10))
Site_CVR=float(np.sum(KW_Attributes['Conversions']))/float(np.sum(KW_Attributes['Clicks']))
AdjFactor=((KW_Attributes.loc[MktAdj,'CVR_MKT']-Site_CVR)/Site_CVR)/2
KW_Attributes.loc[MktAdj,'Bid']=(1+AdjFactor)*KW_Attributes.loc[MktAdj,'Bid']

#c) Cap bids at reasonable levels, based on their quality score

Q1Adj=KW_Attributes['Quality score']>7
KW_Attributes.loc[Q1Adj,'Bid']=np.minimum(KW_Attributes.loc[Q1Adj,'Bid'],KW_Attributes.loc[Q1Adj,'Est First Pos. Bid'])
Q2Adj=np.logical_and(KW_Attributes['Quality score']>5,KW_Attributes['Quality score']<8)
KW_Attributes.loc[Q2Adj,'Bid']=np.minimum(KW_Attributes.loc[Q2Adj,'Bid'],((0.5*KW_Attributes.loc[Q2Adj,'Est First Pos. Bid'])+(0.5*KW_Attributes.loc[Q2Adj,'Est Top of Page Bid'])))
Q3Adj=KW_Attributes['Quality score']<6
KW_Attributes.loc[Q3Adj,'Bid']=np.minimum(KW_Attributes.loc[Q3Adj,'Bid'],((0.1*KW_Attributes.loc[Q3Adj,'Est First Pos. Bid'])+(0.9*KW_Attributes.loc[Q3Adj,'Est Top of Page Bid'])))

KW_Attributes.loc[KW_Attributes['Bid']>12,'Bid']=12

# Cap bids of broad match KWs

def broad_match_bid_cap(AdGroup):
    """Returns bid cap values for each adgroups, for capping broad match keywords""" 
    Subset=AdGroup[AdGroup['Match type'] == "Exact"]
    return pd.DataFrame({'Bid Cap':np.amin(Subset['Bid'])},index=AdGroup.index)

KW_Attributes['Broad Match Bid Cap']=KW_Attributes.groupby('Ad group').apply(broad_match_bid_cap)

BroadAdj=KW_Attributes['Match type']=="Broad"
KW_Attributes.loc[BroadAdj,'Bid']=np.minimum(KW_Attributes.loc[BroadAdj,'Bid'],KW_Attributes.loc[BroadAdj,'Broad Match Bid Cap'])


#EXPORT BID UPLOAD FILE

#KW_Attributes[['Bid']].to_csv("Bid Upload Python Output.csv")

KW_Attributes.to_csv("Bid Upload Python Output.csv")

File: KeyWordBidding.R
#IMPORT NECESSARY PACKAGES

library(data.table)
library(dplyr)
library(tidyr)
library(stringr)

#IMPORT DATA

Inventory_Current_Onsite=fread("Inventory_Current_Onsite.csv")
Inventory_Historical=fread("Inventory_Historical.csv")
KW_Attributes=fread("KW_Attributes.csv")
KW_Performance_L120D=fread("KW_Performance_L120D.csv")
Make_Model_ASR=fread("Make_Model_ASR.csv")

#INSPECT DATA

str(Inventory_Current_Onsite)
str(Inventory_Historical)
str(KW_Attributes)
str(KW_Performance_L120D)
str(Make_Model_ASR)
summary(Inventory_Current_Onsite)
summary(Inventory_Historical)
summary(KW_Attributes)
summary(KW_Performance_L120D)
summary(Make_Model_ASR)

#CLEANING NOTES:

#No missing values observed
#One empty column in Inventory_Historical
#Few empty trailing rows in Make_Model_ASR
#Following columns need to be converted to integer/numeric type after removing charecters like "$" and ",":
#Est First Pos. Bid (KW_Attributes)
#Est Top of Page Bid (KW_Attributes)
#Cost (KW_Performance_L120D)
#Impressions (KW_Performance_L120D)
#Clicks (KW_Performance_L120D)
#ASR (Make_Model_ASR)

#CLEAN DATA

Inventory_Historical$V5=NULL   #Delete empty column
KW_Performance_L120D$Clicks=as.integer((gsub(",", "", KW_Performance_L120D$Clicks))) # Convert Clicks to integer after removing ","
KW_Performance_L120D$Impressions=as.integer((gsub(",", "", KW_Performance_L120D$Impressions))) 
KW_Performance_L120D$Cost=as.numeric((gsub("\\$|,", "", KW_Performance_L120D$Cost))) # Convert to numeric after removing "$" sign and/or ","
Make_Model_ASR=Make_Model_ASR[!apply(Make_Model_ASR == "", 1, all),] # Remove empty rows
Make_Model_ASR$ASR=as.numeric((gsub("\\$|,", "", Make_Model_ASR$ASR))) 
KW_Attributes$`Est First Pos. Bid`= as.numeric((gsub("\\$|,", "", KW_Attributes$`Est First Pos. Bid`))) 
KW_Attributes$`Est Top of Page Bid`= as.numeric((gsub("\\$|,", "", KW_Attributes$`Est Top of Page Bid`))) 


#EXTRACT ATTRIBUTES (MKT, MAKE, MODEL and YEAR FROM AD GROUP)

KW_Attributes$MKT= str_extract(KW_Attributes$`Ad group`,"(?<=SRCH-I-)(.*)(?=-MK_)")# Extract MKT from Ad group
KW_Attributes$Make=str_extract(KW_Attributes$`Ad group`,"(?<=-MK_)(.*)(?=-MO_)")
KW_Attributes$Model=str_extract(KW_Attributes$`Ad group`,"(?<=-MO_)(.*)(?=-YR_)")
KW_Attributes$Year=as.integer(paste0("20",str_extract(KW_Attributes$`Ad group`,"(?<=-YR_)(.*)")))

#DATA INTEGRITY CHECKS

print(sum(duplicated(KW_Attributes$`KW ID`))) #prints 0 if no duplictes found
print(sum(duplicated(KW_Performance_L120D$`KW ID`)))
print(sum(duplicated(Make_Model_ASR$`Make Model`)))
print(sum(duplicated(paste0(Inventory_Current_Onsite$Make,Inventory_Current_Onsite$Model,Inventory_Current_Onsite$Year))))
print(sum(duplicated(paste0(Inventory_Historical$Make,Inventory_Historical$Model,Inventory_Historical$Year))))
print(anti_join(KW_Attributes,KW_Performance_L120D,by="KW ID"))# prints 0 rows if all the rows in KW_Attributes have a match in KW_Performance_L120D by KW ID
print(anti_join(KW_Performance_L120D,KW_Attributes,by="KW ID"))
print(anti_join(KW_Attributes,Inventory_Current_Onsite,by=c("Make","Model","Year")))
print(anti_join(Inventory_Current_Onsite,KW_Attributes,by=c("Make","Model","Year")))
print(anti_join(KW_Attributes,Inventory_Historical,by=c("Make","Model","Year")))
print(anti_join(Inventory_Historical,KW_Attributes,by=c("Make","Model","Year")))
print(anti_join(KW_Attributes,Make_Model_ASR,by=c("Make","Model")))
print(anti_join(Make_Model_ASR,KW_Attributes,by=c("Make","Model")))

#DATA INTEGRITY CHECKS - SUMMARY:
#No duplicates found in unique keys (KW ID) and composite unique keys ("Make","Model","Year" & "Make","Model")
#No referential integrity violations observed


#PREPARE DATA

setkey(KW_Attributes,`KW ID`)
setkey(KW_Performance_L120D,`KW ID`)
KW_Attributes=KW_Attributes[KW_Performance_L120D, nomatch=0] #Join KW_Attributes with KW_Performance_L120D (the data.table way) for fetching performance metrics
rm(KW_Performance_L120D) # remove KW_Performance_L120D as its now redundant
KW_Attributes[,Conversions_Ad_group := sum(Conversions), 
                                                   by = .(`Ad group`)][,Conversions_Make_Model_Year := sum(Conversions),
                                                                       by = .(Make,Model,Year)][,Conversions_Make_Model := sum(Conversions),by = .(Make,Model)]
KW_Attributes[,CVR := Conversions/Clicks][,CVR_Ad_group := sum(Conversions)/sum(Clicks), 
                                                             by = .(`Ad group`)][,CVR_Make_Model_Year := sum(Conversions)/sum(Clicks),
                                                                                 by = .(Make,Model,Year)][,CVR_Make_Model := sum(Conversions)/sum(Clicks),
                                                                                                          by = .(Make,Model)][,CVR_MKT := sum(Conversions)/sum(Clicks),by = .(MKT)]
KW_Attributes=KW_Attributes%>%inner_join(Make_Model_ASR,by=c("Make","Model"))%>%select(-`Make Model`)
rm(Make_Model_ASR)

#STEP 1- CALCULATE INITIAL BID


for (row in 1:nrow(KW_Attributes)) {
  
  if(KW_Attributes[row, "Conversions"] > 10) {
    
    KW_Attributes[row, "Bid"]=KW_Attributes[row, "CVR"]*KW_Attributes[row, "ASR"]
  
    }else if(KW_Attributes[row, "Conversions_Ad_group"] > 10){
    
    KW_Attributes[row, "Bid"]=KW_Attributes[row, "CVR_Ad_group"]*KW_Attributes[row, "ASR"]   
    }else if(KW_Attributes[row, "Conversions_Make_Model_Year"] > 10){
    
    KW_Attributes[row, "Bid"]=KW_Attributes[row, "CVR_Make_Model_Year"]*KW_Attributes[row, "ASR"]   
    }else if(KW_Attributes[row, "Conversions_Make_Model"] > 10){
    
    KW_Attributes[row, "Bid"]=KW_Attributes[row, "CVR_Make_Model"]*KW_Attributes[row, "ASR"]   
    }else{
    
    KW_Attributes[row, "Bid"]=KW_Attributes[row, "Est First Pos. Bid"]
    
    }  
    
}


#STEP 2- ADJUST CALCULATED BID


KW_Attributes=KW_Attributes%>%inner_join(Inventory_Current_Onsite,by=c("Make","Model","Year"))%>%
  inner_join(Inventory_Historical,by=c("Make","Model","Year"))

rm(Inventory_Current_Onsite)
rm(Inventory_Historical)

for (row in 1:nrow(KW_Attributes)) {
  
  
  
  if(KW_Attributes[row, "CurrentOnsiteInventory"] < KW_Attributes[row, "HistAvgInv"]) {
    
    AdjFactor=((KW_Attributes[row, "HistAvgInv"]-KW_Attributes[row, "CurrentOnsiteInventory"])/KW_Attributes[row, "HistAvgInv"])/2
    KW_Attributes[row, "Bid"]=(1-AdjFactor)*KW_Attributes[row, "Bid"]
    
  }
  if(KW_Attributes[row, "Conversions_Ad_group"] < 11){
    
    Site_CVR=sum(KW_Attributes$Conversions)/sum(KW_Attributes$Clicks)
    AdjFactor=((KW_Attributes[row, "CVR_MKT"]-Site_CVR)/Site_CVR)/2
    KW_Attributes[row, "Bid"]=(1+AdjFactor)*KW_Attributes[row, "Bid"]
    rm(AdjFactor)
    rm(Site_CVR)
  }
  if(KW_Attributes[row, "Quality score"] > 7){
    
    KW_Attributes[row, "Bid"]=min(KW_Attributes[row, "Bid"],KW_Attributes[row, "Est First Pos. Bid"])
    
  }else if(KW_Attributes[row, "Quality score"] > 5){
    
    KW_Attributes[row, "Bid"]=min(KW_Attributes[row, "Bid"],(0.5*KW_Attributes[row, "Est First Pos. Bid"]+0.5*KW_Attributes[row, "Est Top of Page Bid"]))
  }else{
    
    KW_Attributes[row, "Bid"]=min(KW_Attributes[row, "Bid"],(0.1*KW_Attributes[row, "Est First Pos. Bid"]+0.9*KW_Attributes[row, "Est Top of Page Bid"]))
    
  } 
  if(KW_Attributes[row, "Bid"]>12)
  {
    KW_Attributes[row, "Bid"]=12
  }

  
}


Temp = KW_Attributes%>% # Bid cap for Broad match type
  filter(`Match type` == 'Exact')%>%
  group_by(`Ad group`)%>%
  summarise(BidCap=min(Bid))

KW_Attributes=KW_Attributes%>%
  inner_join(Temp,by=("Ad group"))

for (row in 1:nrow(KW_Attributes)) {

if(KW_Attributes[row, "Match type"]=="Broad")
{
  KW_Attributes[row, "Bid"]=min(KW_Attributes[row, "Bid"],KW_Attributes[row, "BidCap"])  
}
  
}  

Bid_upload=KW_Attributes[,c("KW ID","Bid")]

rm(KW_Attributes)
rm(Temp)

#EXPORT BID UPLOAD FILE

write.csv(Bid_upload,"Bid Upload R Output.csv")








